receiver/googlecloudspannerreceiver/internal/filterfactory/filterbymetricbuilder.go

```diff
@@ -0,0 +1,154 @@
+}
+
+func (b filterByMetricBuilder) handleLowCardinalityGroups(groups []*metadata.MetricsMetadata, remainingTotalLimit int,
+	filterByMetric map[string]filter.ItemFilter) (int, error) {
+
+	if len(groups) == 0 {
+		return remainingTotalLimit, nil
+	}
+
+	limitPerMetricByTimestamp := b.config.ProjectAmount * b.config.InstanceAmount * b.config.DatabaseAmount
+	// For low cardinality metrics total limit is equal to limit by timestamp
+	totalLimitPerMetric := limitPerMetricByTimestamp
+
+	b.logger.Debug("Calculated cardinality limits for low cardinality metric group",
+		zap.Int("limitPerMetricByTimestamp", limitPerMetricByTimestamp),
+		zap.Int("totalLimitPerMetric", totalLimitPerMetric))
+
+	return b.constructFiltersForGroups(totalLimitPerMetric, limitPerMetricByTimestamp, groups, remainingTotalLimit, filterByMetric)
+}
+
+func (b filterByMetricBuilder) handleHighCardinalityGroups(groups []*metadata.MetricsMetadata, remainingTotalLimit int,
+	filterByMetric map[string]filter.ItemFilter) (int, error) {
+
+	if len(groups) == 0 {
+		return remainingTotalLimit, nil
+	}
+
+	totalLimitPerMetric := remainingTotalLimit / totalAmountOfMetricsInGroups(groups)
+	limitPerMetricByTimestamp := totalLimitPerMetric / defaultMetricDataPointsAmountInPeriod
+
+	b.logger.Debug("Calculated cardinality limits for high cardinality metric group",
+		zap.Int("limitPerMetricByTimestamp", limitPerMetricByTimestamp),
+		zap.Int("totalLimitPerMetric", totalLimitPerMetric))
+
+	if limitPerMetricByTimestamp < 1 {
+		return remainingTotalLimit, errors.New("limit per metric per timestamp for high cardinality metrics is lower than 1")
+	}
+
+	return b.constructFiltersForGroups(totalLimitPerMetric, limitPerMetricByTimestamp, groups, remainingTotalLimit, filterByMetric)
+}
+
+func (b filterByMetricBuilder) constructFiltersForGroups(totalLimitPerMetric int, limitPerMetricByTimestamp int,
+	groups []*metadata.MetricsMetadata, remainingTotalLimit int, filterByMetric map[string]filter.ItemFilter) (int, error) {
+
+	newTotalLimit := remainingTotalLimit
+
+	for _, metadataItem := range groups {
+		for _, metricValueMetadata := range metadataItem.QueryMetricValuesMetadata {
+			newTotalLimit -= totalLimitPerMetric
+			metricFullName := metadataItem.MetricNamePrefix + metricValueMetadata.Name()
+
+			b.logger.Debug("Setting cardinality limits for metric",
+				zap.String("metricFullName", metricFullName),
+				zap.Int("limitPerMetricByTimestamp", limitPerMetricByTimestamp),
+				zap.Int("totalLimitPerMetric", totalLimitPerMetric),
+				zap.Int("remainingTotalLimit", newTotalLimit))
+
+			itemFilter, err := filter.NewItemCardinalityFilter(metricFullName, totalLimitPerMetric,
+				limitPerMetricByTimestamp, defaultItemActivityPeriod, b.logger)
+			if err != nil {
+				return remainingTotalLimit, err
+			}
+			filterByMetric[metricFullName] = itemFilter
+		}
+	}
+
+	return newTotalLimit, nil
+}
+
+func totalAmountOfMetricsInGroups(metadataItems []*metadata.MetricsMetadata) int {
```

Comment:
```suggestion
func countMetricsInGroups(metadataItems []*metadata.MetricsMetadata) (amount int) 
```

Removes the need for declaring a variable and keeps the intent of the function clear.

Reply:
Agree

Reply:
Done.